### PR TITLE
[Bugfix][NPU] Materialize causal mask in FlashAttention dense path

### LINKED ISSUE #7322: [Bug]: NPU FlashAttention ignores causal=True in the dense attention path
### Your current environment

  - Hardware: Ascend 950PR, single NPU
  - Container: `quay.io/ascend/vllm-omni:v0.28.0-a5`
  - Attention backend: `FLASH_ATTN`
  - Model: `nvidia/Cosmos3-Nano`

### Your code version

  - vLLM-Omni base: `v0.28.0-101-g087ac463`
    - Base commit: `087ac46318c42d1fa1e4de13a8053f94683898bd`
  - vLLM: `0.28.0+empty`
    - Source commit: `2cf0a6915ce544dc493a0990f2ea38d81601128a`
  - vLLM-Ascend: package metadata reports `0.0.0`
    - Source checkout is provided by the container image
    - Exact source commit is unavailable because the image does not retain its `.git` directory
  - PyTorch: `2.10.0+cpu`
  - torch-npu: `2.10.0.post4`
  - MindIE-SD: `3.1.0`
  - diffusers: `0.40.0`
  - transformers: `5.14.1`

### 🐛 Describe the bug

  On Ascend NPU, FlashAttentionImpl.forward_fa_npu() does not apply the
  causal constraint when the attention layer is constructed with
  causal=True.

  The CUDA FlashAttention paths forward self.causal to the FlashAttention
  kernel. However, the NPU dense path only forwards attn_mask to
  mindiesd.attention_forward():

  attention_mask = _maybe_reshape_attn_mask(
      query,
      key,
      attention_mask,
      mask_mode="full_qk",
  )

  return attention_forward(
      query,
      key,
      value,
      attn_mask=attention_mask,
      ...
  )

  When attn_metadata does not contain an explicit mask,
  attention_mask remains None, even if self.causal is True.
  The NPU attention is therefore executed as unmasked bidirectional
  attention.

  Cosmos3 exposes this problem because its understanding pathway constructs
  the self-attention layer with causal=True.

  ### Reproduction

  Start Cosmos3-Nano on one Ascend 950PR NPU:

  export ASCEND_RT_VISIBLE_DEVICES=<device-id>
  export VLLM_OMNI_LOGGING_LEVEL=DEBUG

  vllm serve <COSMOS3_NANO_PATH> \
    --served-model-name "nvidia/Cosmos3-Nano" \
    --omni \
    --host 0.0.0.0 \
    --port 8000 \
    --init-timeout 1800 \
    --model-class-name Cosmos3OmniDiffusersPipeline \
    --no-guardrails \
    --allowed-local-media-path / \
    --enforce-eager \
    --max-model-len 8192 \
    --tensor-parallel-size 1 \
    --vae-use-tiling \
    --diffusion-attention-backend FLASH_ATTN

  Send a text-to-video request:

  curl -X POST http://127.0.0.1:8000/v1/videos/sync \
    -H "Accept: video/mp4" \
    -F 'model=nvidia/Cosmos3-Nano' \
    -F 'prompt=A robot arm is cleaning a plate in the kitchen' \
    -F 'negative_prompt=blurry, distorted, low quality, jittery, deformed' \
    -F 'size=1280x720' \
    -F 'num_frames=25' \
    -F 'fps=24' \
    -F 'num_inference_steps=35' \
    -F 'guidance_scale=6.0' \
    -F 'max_sequence_length=4096' \
    -F 'flow_shift=10.0' \
    -F 'seed=123' \
    -o output.mp4

  ### Actual behavior

  The request completes and returns a video, but the generated frames are
  semantically meaningless noise.

  Instrumenting or mocking mindiesd.attention_forward() shows that
  attn_mask=None is passed even though the attention implementation has
  self.causal=True.

  No exception is raised, which makes this a silent correctness issue.

  ### Expected behavior

  The NPU backend should preserve the causal semantics requested by the model.

  For query length Sq and key length Skv, the mask should use the same
  bottom-right alignment as FlashAttention:

  key_position <= query_position + (Skv - Sq)

  If an explicit keep-mask is also supplied, the effective mask should be the
  logical intersection of the explicit mask and the causal mask.

  Non-causal attention should remain unchanged.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Oliver7th**
## Purpose

  Fixes #7322 

  The Ascend NPU implementation of
  `FlashAttentionImpl.forward_fa_npu()` did not preserve the causal
  semantics requested by the model.

  Unlike the CUDA FlashAttention path, the dense MindIE-SD NPU path does
  not receive a `causal` argument. Before this change, when
  `self.causal=True` and no explicit attention mask was provided,
  `mindiesd.attention_forward()` was called with `attn_mask=None`.
  Consequently, causal self-attention was silently executed as full
  bidirectional attention.

  Cosmos3-Nano exposed this issue in its understanding pathway. In the
  affected configuration, inference completed without an exception, but
  the generated video contained semantically meaningless noise.

  This PR:

  1. Materializes a boolean causal keep-mask when `self.causal=True`.
  2. Uses FlashAttention-compatible bottom-right alignment when
     `Sq != Skv`.
  3. Combines the causal mask with an existing explicit keep-mask using
     `torch.logical_and`.
  4. Makes the effective mask contiguous before passing it to MindIE-SD.
  5. Preserves the existing unmasked behavior when `self.causal=False`.

  The effective causal condition is:

  ```python
  key_position <= query_position + (key_length - query_length)
  ```

  The change is limited to the dense NPU FlashAttention path.

  ## Test Plan

  ### Unit tests

  The tests in `tests/diffusion/attention/test_flash_attn.py` cover:

  - causal attention with `Sq == Skv`
  - bottom-right alignment with `Sq < Skv`
  - bottom-right alignment with `Sq > Skv`
  - composition with an explicit keep-mask
  - composition with the NPU variable-length fallback padding mask
  - both default and `ascend_laser_attention` fallback modes
  - preservation of `attn_mask=None` for non-causal attention

  Command:

  ```bash
  pytest -q tests/diffusion/attention/test_flash_attn.py \
    -k "npu_causal or npu_noncausal_without_explicit_mask"
  ```

  ### End-to-end validation

  Hardware and runtime:

  - Ascend 950PR
  - Single-NPU inference
  - Container: `quay.io/ascend/vllm-omni:v0.28.0-a5`
  - CANN Toolkit: 9.1.0
  - torch-npu: 2.10.0.post4
  - MindIE-SD: 3.1.0
  - Model: `nvidia/Cosmos3-Nano`
  - Attention backend: `FLASH_ATTN`

  Server command:

  ```bash
  vllm serve <COSMOS3_NANO_PATH> \
    --served-model-name "nvidia/Cosmos3-Nano" \
    --omni \
    --host 0.0.0.0 \
    --port 8000 \
    --init-timeout 1800 \
    --model-class-name Cosmos3OmniDiffusersPipeline \
    --no-guardrails \
    --allowed-local-media-path / \
    --enforce-eager \
    --max-model-len 8192 \
    --tensor-parallel-size 1 \
    --vae-use-tiling \
    --diffusion-attention-backend FLASH_ATTN
  ```

  T2V case:

  - Prompt: `A robot arm is cleaning a plate in the kitchen`
  - Negative prompt: `blurry, distorted, low quality, jittery, deformed`
  - Resolution: `1280x720`
  - Frames: 25
  - FPS: 24
  - Inference steps: 35
  - Guidance scale: 6.0
  - Seed: 123

  ## vLLM Version

  - Version: `0.28.0+empty`
  - Commit: `2cf0a6915ce544dc493a0990f2ea38d81601128a`

  ## vLLM-Omni Commit

  - Base commit:
    `087ac46318c42d1fa1e4de13a8053f94683898bd`
  - PR HEAD: `729cf6a751a5037eebe0540db2edbdb75784f1cd`

  ## Test Result

  - All newly added causal-mask test assertions passed.
  - Before the fix, the Cosmos3-Nano request completed but generated
    semantically meaningless noise.
  - After the fix, the same request returned HTTP 200 and generated a
    semantically valid 25-frame video.
  - Non-causal NPU attention remains unmasked.
  - Existing explicit masks are preserved and intersected with the causal
    constraint.

  ### Environment-specific note

  The A5 development image may report:

  ```text
  corrupted size vs. prev_size
  ```

  during Python interpreter or server teardown. This occurs after successful
  inference and artifact generation and was also reproduced on the unmodified
  base environment. It is unrelated to the causal-mask change.

### CODE REVIEW

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 1]
```diff
@@ -588,6 +588,20 @@ def forward_fa_npu(
...
 
+        if self.causal:
+            # MindIE-SD's dense NPU path has no causal flag, so encode the
+            # constraint in an explicit boolean keep-mask (True means attend).
+            # Match FlashAttention's bottom-right alignment when Sq != Skv.
+            causal_mask = torch.ones(
+                (query.shape[1], key.shape[1]),
+                dtype=torch.bool,
+                device=query.device,
+            ).tril(diagonal=key.shape[1] - query.shape[1])
```
Comment (david6666666) [inline vllm_omni/diffusion/attention/backends/flash_attn.py]: For Sq=4, Skv=2, the first two rows are fully masked. MindIE forwards this to npu_fusion_attention with the default inner_precise=0; Ascend documents inner_precise=2 for handling fully masked rows. Please ensure these rows return zero, as FlashAttention requires, and add an NPU output test—the current mock only checks the mask. See [the operator contract](https://github.com/Ascend/op-plugin/blob/26.0.0/docs/zh/custom_APIs/torch_npu/torch_npu-npu_fusion_attention.md).

Comment (Oliver7th) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: Thanks for pointing out the fully masked row case.

I revised the implementation to use the native Ascend causal-attention mode instead of materializing the bottom-right causal mask and forwarding it through the generic MindIE-SD dense-attention interface.

The reason for the lower-level compatibility fallback is that the current mindiesd.attention_forward interface does not fully expose causal-attention semantics. In particular:

- It does not provide a causal argument.
- Its dense fused_attn_score path forwards the mask to torch_npu.npu_fusion_attention with the default sparse_mode=0 and inner_precise=0.
- Additional sparse_mode or inner_precise kwargs passed to attention_forward are not propagated to the underlying operator.

Consequently, the existing MindIE-SD interface cannot express FlashAttention's right-down/bottom-right causal contract or explicitly control invalid-row handling.

To keep the workaround narrow:
- causal=True uses torch_npu.npu_fusion_attention directly with sparse_mode=3.
- Sq > Skv additionally uses inner_precise=2.
- Non-causal attention continues to use the original MindIE-SD path unchanged.

This direct call is intended as a compatibility fallback until MindIE-SD exposes complete causal-attention controls. It still uses the same underlying torch_npu.npu_fusion_attention operator used by MindIE-SD, rather than introducing a separate attention implementation.

I also added a real Ascend NPU output test for Sq=4, Skv=2. It verifies that the first two fully masked rows are exactly zero, all outputs are finite, and the remaining rows match an FP32 bottom-right causal reference.

Validation results:

- Targeted unit tests: 4 passed
- Native Ascend NPU output test: 1 passed
- Cosmos3-Nano 25-frame T2V end-to-end test on Ascend 950PR: passed

Once MindIE-SD provides a causal-aware interface that propagates sparse_mode and inner_precise, this fallback can be replaced with the regular backend call.

Comment (david6666666) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: The native path and output test address the fully masked rows. One issue remains: this early return ignores `attn_metadata.attn_mask`. With Sq=Skv=4 and keep-mask `[True, False, True, True]`, later queries can now attend to the excluded second key. Please keep this fast path for unmasked inputs and preserve causal + explicit-mask composition, including its regression test.

Comment (david6666666) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: Both review points are addressed in 8a8e93be: explicit masks are composed with causality, and fully masked rows enable invalid-row handling. The regression tests cover both cases. No remaining blocking findings from my review; I have not rerun the NPU tests locally.

**Comment by david6666666** [comment]
please paste accuracy and perf comparasion

**Comment by david6666666** [comment]
@fan2956 ptal thx

**Comment by Oliver7th** [comment]
> comparasion
[visual.zip](https://github.com/user-attachments/files/32162292/visual.zip)

**Review by david6666666 — APPROVED** [review]
LGTM now
